keanu-project/src/test/java/io/improbable/keanu/e2e/regression/RadonHierarchicalRegression.java

```diff
@@ -52,134 +50,114 @@ public void readRadonCSV() {
+                                    DoubleVertex eps) {
+
         double[] floorForSubModel = data.stream().mapToDouble(d -> d.floor).toArray();
         double[] radonForSubModel = data.stream().mapToDouble(d -> d.log_radon).toArray();
 
-        DoubleTensor x = DoubleTensor.create(floorForSubModel, floorForSubModel.length, 1);
-        DoubleTensor y = DoubleTensor.create(radonForSubModel, floorForSubModel.length, 1);
+        DoubleVertex x = ConstantVertex.of(DoubleTensor.create(floorForSubModel, floorForSubModel.length));
+
+        DoubleVertex gradient = new GaussianVertex(muGradient, sigmaGradient);
+        DoubleVertex intercept = new GaussianVertex(muIntercept, sigmaIntercept);
 
-        return RegressionModel.
-            withTrainingData(x, y).
-            withRegularization(RegressionRegularization.RIDGE).
-            withPriorOnWeights(muGradient, sigmaGradient).
-            withPriorOnIntercept(muIntercept, sigmaIntercept).
-            buildWithoutFitting();
+        DoubleVertex y = x.times(gradient).plus(intercept);
+
+        DoubleVertex yObs = new GaussianVertex(y, eps);
+        yObs.observe(DoubleTensor.create(radonForSubModel, floorForSubModel.length));
+
+        return new SubModel(gradient, intercept);
     }
 
-    private NetworkSamples NUTSSample(BayesianNetwork bayesianNetwork) {
-        Vertex muAlpha = bayesianNetwork.getVertexByLabel(new VertexLabel("MuIntercept"));
-        Vertex sigmaAlpha = bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaIntercept"));
-        Vertex muBeta = bayesianNetwork.getVertexByLabel(new VertexLabel("MuGradient"));
-        Vertex sigmaBeta = bayesianNetwork.getVertexByLabel(new VertexLabel("SigmaGradient"));
+    private NetworkSamples sampleWithNUTS(BayesianNetwork bayesianNetwork, List<Vertex> sampleFrom) {
 
-        // note that way too few samples are taken due to time constraints
+        int sampleCount = 1000;
         KeanuProbabilisticModelWithGradient probabilisticModel = new KeanuProbabilisticModelWithGradient(bayesianNetwork);
-        return Keanu.Sampling.NUTS.builder()
-            .maxTreeHeight(5)
-            .adaptEnabled(true)
+        return NUTS.builder()
+            .adaptCount(sampleCount)
+            .maxTreeHeight(8)
             .build()
-            .getPosteriorSamples(probabilisticModel, Arrays.asList(muAlpha, muBeta, sigmaAlpha, sigmaBeta), 10)
-            .downSample(10);
-    }
-
-    private void optimise(BayesianNetwork bayesianNetwork) {
-        bayesianNetwork.probeForNonZeroProbability(100);
-        GradientOptimizer optimizer = Keanu.Optimizer.Gradient.builderFor(bayesianNetwork)
-            .algorithm(ConjugateGradient.builder()
-                .maxEvaluations(10000)
-                .build()
-            ).build();
-        optimizer.maxAPosteriori();
+            .getPosteriorSamples(probabilisticModel, sampleFrom, sampleCount)
+            .drop(sampleCount / 4);
```

Comment:
It's only ever dependent on the number of latents in the single variable proposal with Metropolis Hastings. That's because you need to change all variables if you want to avoid correlation in the samples. And if you're changing a single variable at a time then it will take N (number of latents) before you've changed them all. Even then that's only for down sampling. The number of samples you should drop depends on how many samples you need to take before your samples show little autocorrelation (don't show any general trend in any direction). I've gone for 25% just to be sure that we're not including any non-converged samples in our average.

Reply:
Cool